cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/kafkaassigner/KafkaAssignerDiskUsageDistributionGoal.java

```diff
@@ -0,0 +1,568 @@
+   */
+  boolean swapReplicas(Broker toSwap, 
+                       Broker toSwapWith, 
+                       double meanDiskUsage, 
+                       ClusterModel clusterModel,
+                       Set<String> excludedTopics) throws AnalysisInputException {
+    LOG.trace("Swapping replicas between broker {}({}) and broker {}({})",
+             toSwap.id(), brokerSize(toSwap), toSwapWith.id(), brokerSize(toSwapWith));
+    double sizeToChange = toSwap.capacityFor(DISK) * meanDiskUsage - brokerSize(toSwap);
+    List<ReplicaWrapper> sortedReplicasToSwap = sortReplicasAscend(toSwap.replicas(), excludedTopics);
+    List<ReplicaWrapper> sortedLeadersToSwapWith = sortReplicasAscend(toSwapWith.leaderReplicas(), excludedTopics);
+    List<ReplicaWrapper> sortedFollowersToSwapWith = sortReplicasAscend(followerReplicas(toSwapWith), excludedTopics);
+    
+    int startPos;
+    int delta;
+    if (sizeToChange > 0) {
+      // iterate from small replicas to large replicas.
+      startPos = 0;
+      delta = 1;
+    } else {
+      // iterate from large replicas to small replicas.
+      startPos = sortedReplicasToSwap.size() - 1;
+      delta = -1;
+    }
+    
+    for (int i = startPos; i >= 0 && i < sortedReplicasToSwap.size(); i += delta) {
+      Replica replicaToSwap = sortedReplicasToSwap.get(i).replica();
+      if (excludedTopics.contains(replicaToSwap.topicPartition().topic())) {
+        continue;
+      }
+      List<ReplicaWrapper> sortedReplicasToSwapWith = 
+          replicaToSwap.isLeader() ? sortedLeadersToSwapWith : sortedFollowersToSwapWith;
+      double sizeToSwap = replicaSize(replicaToSwap);
+      // No need to continue if we are trying to reduce the size and the replicas to swap out is of size 0.
+      if (sizeToChange < 0 && sizeToSwap == 0) {
+        break;
+      }
+      // when sizeToChange > 0, the broker toSwap needs more disk utilization, the replicaToSwapWith should meet the 
+      // following requirements:
+      // 1. replicaToSwapWith.size() > replicaToSwap.size()
+      // 2. After the swap, the disk usage of broker toSwap should not be more than the disk usage of broker
+      //    toSwapWith before the swap.
+      // 3. After the swap, the disk usage of broker toSwapWith should not be less than the disk usage of broker 
+      //    toSwap before the swap.
+      //
+      // When sizeToChange < 0, the broker toSwap needs less disk utilization, the replicaToSwapWith should meet the
+      // following requirements:
+      // 4. replicaToSwapWith.size < replicaToSwap.size()
+      // 5. After the swap, the disk usage of broker toSwap should not be less than the disk usage of broker
+      //    toSwapWith before the swap.
+      // 6. After the swap, the disk usage of broker toSwapWith should not be more than the disk usage of broker 
+      //    toSwap before the swap.
+      //
+      // We do not require the swap to be under the balance upper limit or lower limit. Instead, we just ensure
```

Comment:
Do we assume that each broker has equal disk capacity? Should we note that in these comments?

Reply:
No, we do not assume that. The upper limit and lower limit are usage in percentage.